### ISSUE TITLE
Built-in TLS Decryption

### ISSUE THREAD

**Issue by emanuele-f** [emanuele-f/PCAPdroid#57]
At the current status, PCAPdroid requires an external proxy in order to decrypt TLS packets. On the other hand, a built-in solution would provide the following advantages:

- No need to have a PC to decrypt TLS
- PCAPdroid could show the decrypted payload directly into the app. This is related to #38.
- PCAPdroid could produce a ~decrypted~ decryptable PCAP

## Requirements

- Should support TLS 1.3
- Should not impact the main PCAPdroid development. TLS decryption is a secondary topic in PCAPdroid, so this feature should be supported by an active third party
- Should comply to the [F-Droid policy](https://f-droid.org/en/docs/Inclusion_Policy) (e.g. do not bundle any binary)
- The solution should be light, primary in terms of device resources (cpu, memory) and possibly in terms of dependencies and APK size

## Candidates

- [LittleProxy-mitm](https://github.com/koh-osug/LittleProxy-mitm)/[LittleProxy](https://github.com/mrog/LittleProxy) is an HTTP proxy. It fits the purpose of a ready-to-use library, however depends on [netty](https://github.com/netty/netty), which is big dependency.
- [nitmproxy](https://github.com/chhsiao90/nitmproxy) is a self contained SOCKS/HTTP proxy also using netty. SOCKS offers the possibility to decrypt non-HTTPS traffic.

Other approaches:
- [NetBare](https://github.com/MegatronKing/NetBare): unmaintained, requires heavy modifications to only extract the proxy code, does not support TLS 1.3, it has the advantage of using `java.nio` for networking, which fits the minimal dependencies requirement
- any native C/C++ library could be integrated

## User Experience

Users will have an option to enable the built-in TLS decryption by a toggle. The old mitmproxy option will be replaced with a more generic ability to use socks5 proxy, which is independent from the TLS decryption. Once enabled, the TLS decryption will happen under the hoods; android proxy settings will be untouched. It will have the following effects:

- In the connections view, the TLS connections data will be replaced by decrypted protocol data, e.g. HTTPS will be decrypted and the HTTP request/reply will be shown
- ~The PCAP will also contain the decrypted payload data in place of the original TLS data~ The traffic could be dumped in the PCAPNG file format, which contains a "Decryption Secrets Block" to store the decryption keys.

In essence TLS will be decapsulated and its inner data shown.

## Roadmap

The built-in decryption is now implemented via the [PCAPdroid-mitm addon](https://github.com/emanuele-f/PCAPdroid-mitm).

- [x] Implement socks5 client in PCAPdroid (#44)
- [x] Build the mitm proxy addon
- [x] Verify that embedded mitm proxy properly proxies the packets received by the socks5 interface
- [x] Show the decrypted connection in PCAPdroid
- [x] Report rejected certificates warnings from mitmproxy in the app
- [x] There are some TLS connections which are not decrypted, possibly fix/report to the user why this happens
- [x] Provide certificate installation via Intent on Android < 11
- [x] Harden the proxy via the SOCKS5 authentication
- [x] Fix export of SSLKEYLOG
- [x] Improve matching of HTTP request/replies in the UI (e.g. highlight row when clicking "jump" icon, or sort connections)
- [x] Documentation, updates docs on methods to bypass untrusted user certs (on Android 7, e.g. magisk trust user certs, bromite chrome://flags),  pinning (LSposed module `sslunpinning`) and Certificate transparency (avoid using a system cert or use magisk hide) - see [the docs](https://emanuele-f.github.io/PCAPdroid/tls_decryption_new)
- [x] Suggest users to disable [Autostart](https://www.vivo.com/en/support/questionByTitle?title=How%20to%20turn%20on/off%20Autostart%20for%20my%20apps) or similar software which can prevent the mitm addon execution

**Comment by emanuele-f** [comment]
With the purpose of minimizing the dependencies, here is a quick overview of the LittleProxy-mitm depedencies (max depth 3):

```
- koh-osug/LittleProxy-mitm
  - bcpkix-jdk15to18
  - bcprov-jdk15to18
  - conscrypt-openjdk-uber
  - commons-io
  - slf4j-log4j12
  - mrog/LittleProxy
    - guava
    - commons-lang3
    - netty-all
    - barchart-udt-bundle (UDT transport protocol)
    - slf4j-api
```

@koh-osug can you please clarify the need for the additional dependencies of LittleProxy-mitm over LittleProxy? E.g. jdk15to18 and conscrypt. Regarding the project structure, I think it's better to keep LittleProxy-mitm in a separate code base, unless it requires a tight integration with PCAPdroid.

**Comment by koh-osug** [comment]
- I haved removed the `commons-io` dependency. 
- `conscrypt-openjdk-uber` must be excluded when using the library under Android, it only required in Java-SE, Android is already using this internally.
- `bcpkix-jdk15to18` and `bcprov-jdk15to18` is needed for creating the MITM certificates. 
- Under Android the `barchart-udt-bundle` dependency must be excluded, it is optional and not supported and not needed. I have patched the latest version of littleproxy to make it optional under Android with a merge request. 
- I guess, I cannot influence the LittleProxy maintainer to remove `guava` or `commons-lang3`, these 2 libraries are convenient, but usually only a subset of the functions are needed.

I will check with my Android test project how much is left after using the Android shrinker. Usually almost all from the utilitiy libraries and major parts of bouncy castle should get removed resulting in just a few HD floppy disks.

**Comment by emanuele-f** [comment]
Great! I think LittleProxy should be run as a thread in PCAPdroid and communicate with it via the socks5 interface (will I will take care of the socks5 migration asap). This will allow PCAPdroid to carry on the TLS decryption asynchronously while processing the network packets in parallel.

I've added a roadmap for this feature in the issue description above, we can work on a branch until it is complete. You can proceed with the integration. Once the embedded LittleProxy works correctly, I will take care of the final step to show the decrypted data in PCAPdroid/in the PCAP. For this I will need a call API in LittleProxy-mitm to read the decrypted data back in PCAPdroid. The API should expose: the decrypted data `byte[]`, the direction of the data (from/to the internet) and the connection identifier (the 5 tuple, or at least proto, dst_ip, src_port, dst_port)

**Comment by koh-osug** [comment]
I have checked the sizes after shrinking:

![image](https://user-images.githubusercontent.com/5708343/114071602-9876a100-98a1-11eb-9e23-7c4b522debea.png)

- BouncyCastle is just using 261 kb
- guava was reduced to 146 kb
- netty is 812 kb
- LittleProxy-mitm itself is 71 kb

**Comment by koh-osug** [comment]
Filtering the HTTP request seems to be possible with LittleProxy. I hope the filter is the one seeing the plain text. How will you map the plain text to the original requests? I can record the plain text, but somehow then later you have to know which plain text corresponds to which encrypted text.

**Comment by emanuele-f** [comment]
> I have checked the sizes after shrinking

1.3 MB is a good result, I think we can settle with this. Great work!

> Filtering the HTTP request seems to be possible with LittleProxy. I hope the filter is the one seeing the plain text. How will you map the plain text to the original requests? I can record the plain text, but somehow then later you have to know which plain text corresponds to which encrypted text.

The idea is the following: when LittleProxy is started I will register a callback (via java interface). Whenever the proxy decrypts a packet, it will call the callback with the decrypted data and connection information (pre-proxy 5 tuple). With the 5 tuple I can retrieve the original connection and set the plain text data.

In b02989598f93c2bc49dbf8bd645efd39ee85d4fb I've implemented the SOCKS5 proxy client. This replaces the old "tunnel" mode. You should now be able to redirect the TCP packets to LittleProxy by using SOCKS5.

**Comment by koh-osug** [comment]
I will have a look into this. My first concern is if LittleProxy is actually supporting SOCKS5. I'm not familiar with the protocol, but what I see is that it needs a HTTP CONNECT call to trigger the proxy.

**Comment by koh-osug** [comment]
I'm investigating this one as alternative: https://github.com/chhsiao90/nitmproxy
It supports socks and can also be used for SMTP / IMAP with TLS then.

**Comment by koh-osug** [comment]
I had an issue with nitmproxy because the addresses received from SOCKS have been IP addresses, no hostnames anymore. I added a handler to extract the Server Name Indication and this looks good now. I created a [fork](https://github.com/koh-osug/nitmproxy/tree/SNI) and check its stability. LittleProxy might be maintained, but it is focusing too much on HTTP, nitmproxy  supports more protocols and has the MITM feature is included. I have tested in under Linux and by connecting from Android to it.

**Comment by emanuele-f** [comment]
Have you tested it with TLS 1.3 and HTTP2? A SOCKS proxy opens up more possibilities so I also encourage this choice.

**Comment by koh-osug** [comment]
I have tested the page "https://tools.keycdn.com/http2-test" with FF usign the HTTP2 protocol. The web page is working.
I used the web page https://resources.duda.co/tls-1-3 to check for TLS 1.3. This was also working.
Maybe some unit test could be added, too.

**Comment by emanuele-f** [comment]
Great! Please provide a pull request/branch to review when ready! If you want a more interactive technical discussion you can reach me on the telegram group

**Comment by koh-osug** [comment]
Sure, will do. I have an issue with a specific domain. I'm in contact with the developer and I'm analyzing this. It is not a problem in a PC environment, but the same code is not working on Android. Most likely the Android SSLEngine stack is behaving differently.  But the website seems to use a strange combination of TLS protocols, I could not detect this with any other web page.

**Comment by koh-osug** [comment]
The Android internal SSL Engine seems to have some issues with TLS 1.3 and ALPN. I have [patched netty](https://github.com/netty/netty/pull/11157) to be able to use the BouncyCastle TLS implementation which is working for the above mentioned issue. I will wait until this gets accepted to have it available as published dependency.

**Comment by koh-osug** [comment]
The nitmproxy can now transparently detect HTTP/HTTPS and supports also at least Android 8. I have also some older phones which I can check. A new API was added to [plugin loggers]( https://github.com/chhsiao90/nitmproxy/blob/master/src/main/java/com/github/chhsiao90/nitmproxy/listener/HttpListener.java) I will add an AIDL interface around this so that PCAPdroid can call this.

**Comment by emanuele-f** [comment]
The android emulator is a convenient tool to test stuff on different SDK versions and devices. I see that the API is HTTPS oriented, does it provide a way to handle non HTTPS traffic? Also I think it could be better to run the proxy as a thread in PCAPdroid, to save IPC complexity and overhead.

**Comment by kaoh** [comment]
OK, not having it as separate app makes it simpler. Non HTTP is still on the TO DO list.

**Comment by qkmaosjtu** [comment]
Hi!
I am very excited to see that "Built-in TLS Decryption "! 
In addition to supporting TLS1.3, have you considered supporting the quic protocol? It will be very common in the future.

**Comment by koh-osug** [comment]
Whatever is supported by conscrypt using the Java SSL engine can be supported. If this is not on the roadmap of both support will only be possible with external SOCKS based proxy tools supporting quic.

**Comment by qkmaosjtu** [comment]
@emanuele-f Hi emanuele! Any update on this topic?

**Comment by emanuele-f** [comment]
@koh-osug any update on this? Can I help with the tests?

**Comment by emanuele-f** [comment]
PCAPdroid [has minSdkVersion 21](https://github.com/emanuele-f/PCAPdroid/blob/master/app/build.gradle#L10), on which android version the built-in Conscrypt provides ALPN support (links appreciated)? The loggers callbacks can be adapted at any time (e.g. via a fork), that's not a problem.

Ps. Here is the link to for BouncyCastle missing TLS renegotiation support: https://github.com/bcgit/bc-java/issues/593

**Comment by kaoh** [comment]
The APLN support requires a backport from a later Java 9 version to Android (OpenJDK 9+181). The important method is `getApplicationProtocol`. It is still missing in [Android 9](http://aosp.opersys.com/xref/android-9.0.0_r47/xref/libcore/ojluni/src/main/java/javax/net/ssl/SSLEngine.java) but exists in [Android 10](http://aosp.opersys.com/xref/android-10.0.0_r47/xref/libcore/ojluni/src/main/java/javax/net/ssl/SSLEngine.java#1965), i.e. all version below Android 10 have to use a bundled Conscrypt version.

**Comment by emanuele-f** [comment]
Thanks for the clarification. 2 MB overhead is ok, but 8 seems to much for me. Anyway, please note that if BouncyCastle is integrated into PCAPdroid, it must meet the [f-droid policy](https://f-droid.org/en/docs/Inclusion_Policy), so it is either built as a submodule or it is downloaded from a trusted repository.

**Comment by kaoh** [comment]
I have raised a pull request in nitmproxy: https://github.com/chhsiao90/nitmproxy/pull/73 to provide the connection context with the log handler. source/dest IP + port + transmitted data would be sufficient, I assume?

**Comment by emanuele-f** [comment]
I assume that it only handles TCP, so the protocol is not needed. The direction of the data is determined by the callback used (onForwardRequest/onForwardResponse), so we should now have everything needed in the log handler.

**Comment by koh-osug** [comment]
I'm waiting for the next release of it and then will create a pull request draft.

**Comment by kaoh** [comment]
New nitmproxy version was released, I started with the gradle integration and shrinking the apk size.

**Comment by kaoh** [comment]
I added a draft: https://github.com/emanuele-f/PCAPdroid/pull/87 What is the netxt step to integrate the nitmproxy listeners?

**Comment by emanuele-f** [comment]
Thanks! If that's not an issue, please base your branch on `dev` rather than `master`. I will merge dev into master just before the release to allow other users to test it and update the translations. Regarding the integration, I will need any useful info in order to integrate your libraries, e.g. on how to start the proxy and examples on how to integrate with it. If you already know what to modify in PCAPdroid, you can provide a initial integration, otherwise I can integrate it and commit into your branch.

**Comment by kaoh** [comment]
If you can point me to the integration point where the data, IP address, port, etc., is read and written to the pcap I can start to develop a stub.

**Comment by emanuele-f** [comment]
- The proxy should be started/stopped as a thread in [CaptureService.java](https://github.com/emanuele-f/PCAPdroid/blob/dev/app/src/main/java/com/emanuelef/remote_capture/CaptureService.java). You can check how the `mHttpServer` is handled to do the same.
- For the time being, you can direct the packets to the proxy by configuring the SOCK5 proxy into the PCAPdroid settings. Later we will add a dedicate option for this built-in proxy. With you setup you should have a working integration, but without the ability to get the packets payload data.
- The last step is to show the decrypted payload data in PCAPdroid by using the callbacks. We can check this later.

**Comment by emanuele-f** [comment]
Regarding the protect method, I think [BackendChannelBootstrap.java](https://github.com/chhsiao90/nitmproxy/blob/master/src/main/java/com/github/chhsiao90/nitmproxy/channel/BackendChannelBootstrap.java) should be modified to allow the call of an external function call after connection (e.g. protect) as shown in https://stackoverflow.com/questions/43463044/protecting-socket-when-using-netty4-on-android-with-vpnservice

**Comment by kaoh** [comment]
I have added an issue to https://github.com/chhsiao90/nitmproxy/issues/82 for the protect method.

**Comment by emanuele-f** [comment]
Good, it's better to have a generic callback which is called every time a connection is open. In our case we will call protect, other project can use it for other stuff.

**Comment by emanuele-f** [comment]
@kaoh any update on this? It would be great to have this feature in the next PCAPdroid major release.

**Comment by kaoh** [comment]
I'm waiting on https://github.com/chhsiao90/nitmproxy/issues/82 If the socket must be protected this seems to be a requirement, if there is no reaction on this, I could look into this. The problem is that the socket identifier must be a long, but this identifier is not revealed by netty.

**Comment by kaoh** [comment]
The nitmproxy maintainer seems not to handle this. So, I started on my own, I think I'm on the right track now, actually it seems simple, let's see how it works.

**Comment by emanuele-f** [comment]
I'm looking again at this problem as this features is something users are demanding. The Java based libraries evaluated so far are not actively developed and in my opinion not mature enough to carry on this task.

The only tool which I trust can do the job is mitmproxy. In #123 I've implemented a plugin which can pass the HTTP and websockets decrypted data to PCAPdroid. My idea to implement the built-in decryption is to bundle mitmproxy with its dependencies so that PCAPdroid can start/stop it at will and receive the decrypted data. This may either be implemented directly in PCAPdroid or into a separate app. The only obstacle to this is python. It seems like building python for Android is not easy and most projects on this matter are now abandoned.

The most promising one seems to be chaquopy, although closed source. I've start experimenting with it but faced [dependency issues](https://github.com/chaquo/chaquopy/issues/583). Another candidate is [python-for-android](https://github.com/kivy/python-for-android).

In short, I'm willing to implement a very basic app with mitmproxy which can be controlled by PCAPdroid. I'm opting for a separate apk right now both for convenience and because in my tests bundling python with mitmproxy requires about 10 MB of overhead per ABI (so x4) in the apk size.

**Comment by emanuele-f** [comment]
With https://github.com/emanuele-f/PCAPdroid/commit/6f420158731847fc073db9537a44d8bb72d51491 the built-in decryption is now integrated. The decryption requires the installation of the [PCAPdroid-mitm addon](https://github.com/emanuele-f/PCAPdroid-mitm) as a separate apk. I've decided to separate it from the PCAPdroid app because of the size (the addon apk is about ~36 MB) and because it relies on the third-party [chaquopy](https://chaquo.com/chaquopy) gradle plugin, which is closed source and it has build-time dependencies which complicate the build.

This is just an initial integration and there are a lot of things to improve.

**Comment by kaoh** [comment]
These are good news. Is the chaquopy approach raising any security concerns? I.e. because it is closed source, is it safe to used it without any risk of hidden code?

**Comment by emanuele-f** [comment]
> Is the chaquopy approach raising any security concerns? I.e. because it is closed source, is it safe to used it without any risk of hidden code?

These concerns are legit for any closed source software. chaquopy is quite a complicate plugin so it's not easy to review it, but given that it widely known, it has a public github page for issues and a clear monetization plan, I think we can trust it. One drawback though is that the license code cannot be make public, so people building trying to build the `PCAPdroid-mitm` module will run in "demo" mode.

In short, it does its job and it does it well, although it's not open source. Unless other equivalent foss solutions are implemented,  this is what we have now and I'm very satisfied with the result.

**Comment by kenduq** [comment]
So for now, is there any option to produce decryptable or decrypted pcap file for analysis in wireshark? 
I tried to install PCAPdroid-mitm app on my emulator but emulator can not see the app. Install was via adb install .apk

**Comment by joksas** [comment]
Thank you for this useful package!

I have a question which, I guess, is similar to the one by @kenduq. How does one decrypt the PCAP file with the help of the new plugin? I installed the plugin but I couldn't see any changes in the app itself. Are decryption keys saved separately somewhere?

**Comment by emanuele-f** [comment]
> I tried to install PCAPdroid-mitm app on my emulator but emulator can not see the app

This is not released yet. You either need to build PCAPdroid from source (dev branch) or install a pre-built debug apk. Here is a [pre-built apk](https://www.dropbox.com/s/pa8u82s225p9z9r/PCAPdroid_12727bd.apk?dl=1) for the latest commit. Then you should enable the "TLS decryption (preview)" option in the settings.

> How does one decrypt the PCAP file with the help of the new plugin?

This is still a work in progress, at the moment the app shows the decrypted HTTP request data (see https://github.com/emanuele-f/PCAPdroid/issues/123#issuecomment-1030714588). I'm implementing the ability to show the full HTTP request/response in the app

> Are decryption keys saved separately somewhere

I've just added (in 12727bdb248040196543a6c2051d9e4d86c8bb4d) the ability to export the decryption keys, to be used in Wireshark. Be sure to install the latest debug apk (link above), then you can export them (the capture must be running):

![2022-03-09_00-05](https://user-images.githubusercontent.com/5488003/157340899-0cf6b4e3-ce09-413a-bb14-6cb1f62297aa.png)

![2022-03-09_00-05_1](https://user-images.githubusercontent.com/5488003/157340942-8a831fd8-52ac-46c8-8dc8-bb352baacd28.png)

A more practical way to get a decryptable pcap is with #185

**Comment by emanuele-f** [comment]
> Tried to export SSL keylog file but the exported .txt file is empty.

The keylog must be exported *after* the HTTPS connection is established.

The ability to inspect the full HTTP payload from PCAPdroid is now implemented, check out https://github.com/emanuele-f/PCAPdroid/issues/107#issuecomment-1074190167

**Comment by K-techie** [comment]
In latest debug version i can see the app crashing and also hanging when "Full payload" toggle is turned on.

**Comment by emanuele-f** [comment]
> In latest debug version i can see the app crashing and also hanging when "Full payload" toggle is turned on.

Should be fixed in fe26eccf19457ecc84c17ec84e963b04cb571543. [New debug apk](https://www.dropbox.com/s/9o8brpsqme9bws5/PCAPdroid_fd53307.apk?dl=1).

**Comment by NEK-RA** [comment]
Consider also creating whitelist of apps that may be captured.

I guess then the whitelist structure may be like that:

```
package.name, confirmed_version
```
I don't know what's more reliable for version detection - `versionCode` or `versionName`, just because probably average user won't be able to extract `versionCode`, but `versionName` finally just a string, which anybody can fill with anything.

i.e. when loading list of target apps you can mark them:
- everything with targetSdk less than android 7 will be "green" or somehow marked as "decryptable"
- apps which are detected in whitelist and with versionName/versionCode lower/equal to value from whitelist also become "green"
- apps which are detected from whitelist by package name, but with version newer than confirmed in whitelist become "orange", or somehow "may be decryptable"
- others will be "red" or somehow "unconfirmed"

The idea is that amount of directly decryptable apps will be small and finally they won't update each day. So there will be added software confirmed by community, so I'm not sure that this list will be updated very often. I guess also it's worth to make minimal in-app update of this list to once per week or even less (but with ability to manually update).

Also one more notice is to avoid adding to list ROM-specific apps, i.e. something with package `com.android.browser` and etc. Just because depending on ROM it may be completelly different apps

**Comment by besendorf** [comment]
Today I tested the beta TLS decryption with a patched Signal apk and it worked in PCAPdroid. I could see decrypted TLS packets. But in wireshark via the http server and also after saving the pcap in pcapdroid and opening this file afterwards in wireshark I could only see encrypted TLS packets.
Is saving or serving decrypted packages via http not implemented yet?
Is there currently a way to extract the decrypted pcap?

**Comment by emanuele-f** [comment]
> Could you provide a new debug release?

Try [this one](https://www.dropbox.com/s/74mqos7gs33tda1/PCAPdroid_46af0e6.apk?dl=1). The keylog file is now exported when the capture is stopped. If you still experience the crash, please report the stack trace

**Comment by besendorf** [comment]
> > Could you provide a new debug release?
> 
> Try [this one](https://www.dropbox.com/s/74mqos7gs33tda1/PCAPdroid_46af0e6.apk?dl=1). The keylog file is now exported when the capture is stopped. If you still experience the crash, please report the stack trace

Thank you very much and also for the fast response. Worked like a charm.
Might I suggest including the sslkeylog in the pcap that can be saved in the app and also the one streamed via http/udp?
https://wiki.wireshark.org/TLS#embedding-decryption-secrets-in-a-pcapng-file

**Comment by emanuele-f** [comment]
> Thank you very much and also for the fast response. Worked like a charm. Might I suggest including the sslkeylog in the pcap that can be saved in the app and also the one streamed via http/udp?

This will be implemented in #185

**Comment by kasnder** [comment]
This is a great addition to PCAPdroid, and I look forward to using it for my research.

Might it be possible to add an option to drop UDP traffic on port 443? This usually prevents apps from using QUIC, and instead makes these apps fallback to TCP.

There might be some complications with some Private DNS solutions -- in this case, one might just let the user know that this must be disabled?

**Comment by emanuele-f** [comment]
> Might it be possible to add an option to drop UDP traffic on port 443? This usually prevents apps from using QUIC, and instead makes these apps fallback to TCP

I read that this fallback behavior is implemented by chrome. Is this part of the standard or it only applies to some apps? Please provide more info in a separate issue and also possibly an example on how it helps with decryption

> There might be some complications with some Private DNS solutions

Private DNS should user TLS on port 443. Do some of them use UDP on port 443?
